{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gffutils\n",
    "import pandas as pd\n",
    "import os\n",
    "os.chdir(\"/home/tfuruta1/CHM13_changes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "original = gffutils.FeatureDB(\"data/original.db\")\n",
    "mapped = gffutils.FeatureDB(\"data/mapped.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "remove = pd.read_csv(\"changes/1_synonyms_remove.csv\", index_col = 0)\n",
    "add = pd.read_csv(\"changes/1_synonyms_add.csv\", index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KIAA1522</td>\n",
       "      <td>chr1</td>\n",
       "      <td>32601643</td>\n",
       "      <td>32634754</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM183A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43013445</td>\n",
       "      <td>43026887</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LEXM</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54686281</td>\n",
       "      <td>54722525</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FAM102B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108593088</td>\n",
       "      <td>108672314</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PIFO</td>\n",
       "      <td>chr1</td>\n",
       "      <td>111339352</td>\n",
       "      <td>111367700</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>LOC102724428</td>\n",
       "      <td>chr21</td>\n",
       "      <td>41769738</td>\n",
       "      <td>41782385</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>LOC112268293</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40431389</td>\n",
       "      <td>40435898</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>LOC122394733</td>\n",
       "      <td>chrX</td>\n",
       "      <td>102403657</td>\n",
       "      <td>102407006</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>LOC727838</td>\n",
       "      <td>chrX</td>\n",
       "      <td>117450728</td>\n",
       "      <td>117454503</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>LOC389895</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138403603</td>\n",
       "      <td>138404836</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene chromosome      start        end strand         biotype\n",
       "1        KIAA1522       chr1   32601643   32634754      +  protein_coding\n",
       "2         FAM183A       chr1   43013445   43026887      +  protein_coding\n",
       "3            LEXM       chr1   54686281   54722525      +  protein_coding\n",
       "6         FAM102B       chr1  108593088  108672314      +  protein_coding\n",
       "7            PIFO       chr1  111339352  111367700      +  protein_coding\n",
       "..            ...        ...        ...        ...    ...             ...\n",
       "414  LOC102724428      chr21   41769738   41782385      -  protein_coding\n",
       "423  LOC112268293      chr22   40431389   40435898      +  protein_coding\n",
       "435  LOC122394733       chrX  102403657  102407006      +  protein_coding\n",
       "437     LOC727838       chrX  117450728  117454503      +  protein_coding\n",
       "438     LOC389895       chrX  138403603  138404836      +  protein_coding\n",
       "\n",
       "[178 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NHSL3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>32601643</td>\n",
       "      <td>32634754</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CFAP144</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43013445</td>\n",
       "      <td>43026887</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CIMAP2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54686281</td>\n",
       "      <td>54722525</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EEIG2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>108593088</td>\n",
       "      <td>108672314</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CIMAP3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>111339352</td>\n",
       "      <td>111367700</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>SIK1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>41769738</td>\n",
       "      <td>41782385</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>TMA7B</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40431389</td>\n",
       "      <td>40435898</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>TMSB15C</td>\n",
       "      <td>chrX</td>\n",
       "      <td>102403657</td>\n",
       "      <td>102407006</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>CT47C1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>117450728</td>\n",
       "      <td>117454503</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>HAPSTR2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138403603</td>\n",
       "      <td>138404836</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene chromosome      start        end strand         biotype\n",
       "1      NHSL3       chr1   32601643   32634754      +  protein_coding\n",
       "2    CFAP144       chr1   43013445   43026887      +  protein_coding\n",
       "3     CIMAP2       chr1   54686281   54722525      +  protein_coding\n",
       "6      EEIG2       chr1  108593088  108672314      +  protein_coding\n",
       "7     CIMAP3       chr1  111339352  111367700      +  protein_coding\n",
       "..       ...        ...        ...        ...    ...             ...\n",
       "414     SIK1      chr21   41769738   41782385      -  protein_coding\n",
       "423    TMA7B      chr22   40431389   40435898      +  protein_coding\n",
       "435  TMSB15C       chrX  102403657  102407006      +  protein_coding\n",
       "437   CT47C1       chrX  117450728  117454503      +  protein_coding\n",
       "438  HAPSTR2       chrX  138403603  138404836      +  protein_coding\n",
       "\n",
       "[178 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in original.features_of_type(\"gene\"):\n",
    "    \n",
    "    gene_name = (gene.attributes.get(\"gene\") or gene.attributes.get(\"gene_name\"))\n",
    "    if not gene_name:\n",
    "        continue\n",
    "    gene_name = gene_name[0]\n",
    "    match = remove[\n",
    "        (remove[\"gene\"] == gene_name)\n",
    "        & (remove['chromosome'] == gene.seqid)\n",
    "        & (remove['start'] == gene.start\n",
    "        & (remove['end'] == gene.end\n",
    "        & (remove['strand'] == gene.strand\n",
    "    ]\n",
    "    if not match.empty:\n",
    "        original.delete(gene, children = True)\n",
    "        print(f\"Deleted {gene}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62449384"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, row in add.iterrows():\n",
    "    gene_name = row['gene']\n",
    "    chrom = row['chromosome']\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "    strand = row['strand']\n",
    "    for gene in mapped.features_of_type('gene'):\n",
    "        gname = (\n",
    "            gene.attributes.get(\"gene\") or\n",
    "            gene.attributes.get(\"gene_name\")\n",
    "        )\n",
    "        if not gene_name:\n",
    "            continue\n",
    "        gene_name = gene_name[0]\n",
    "        if (\n",
    "            (gname == gene_name)\n",
    "            & (gene.seqid == chrom)\n",
    "            & (gene.start == start)\n",
    "            & (gene.end == end)\n",
    "            & (gene.strand == strand)\n",
    "        ):\n",
    "            original.update([gene])\n",
    "            print(gene)\n",
    "            for descendants in mapped.children(gene, level=None, order_by='start'):\n",
    "                original.update([descendants])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/new_annotation.gff', 'w') as out:\n",
    "    for feature in original.all_features(order_by=('seqid', 'start')):\n",
    "        out.write(str(feature) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
